controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java

```diff
@@ -372,32 +379,33 @@ public boolean checkAuthorization(String resource, AuthHandler.Permissions expec
                 allowedLevel = currentInterceptor.authorize(resource);
             }
             if (allowedLevel.ordinal() < expectedLevel.ordinal()) {
-                return false;
+                throw new AuthorizationException("Access not allowed");
             }
-            return true;
-        } else {
-            return true;
         }
+        return "";
     }
 
-    public boolean checkAuthorizationWithToken(String resource, AuthHandler.Permissions expectedLevel) {
+    private String checkAuthorizationWithToken(String resource, AuthHandler.Permissions expectedLevel) throws AuthorizationException {
         if (isAuthEnabled) {
-            boolean retVal = checkAuthorization(resource, expectedLevel);
-            if (retVal) {
-                PravegaInterceptor interceptor = PravegaInterceptor.INTERCEPTOR_OBJECT.get();
-                interceptor.setDelegationToken(resource, expectedLevel, tokenSigningKey);
-            }
-            return retVal;
-        } else {
-            return true;
+            checkAuthorization(resource, expectedLevel);
+            return createDelegationToken(resource, expectedLevel, tokenSigningKey);
         }
+        return "";
     }
 
-    private String getCurrentDelegationToken() {
+    private String createDelegationToken(String resource, AuthHandler.Permissions expectedLevel, String tokenSigningKey) {
         if (isAuthEnabled) {
-            return PravegaInterceptor.retrieveDelegationToken(tokenSigningKey);
-        } else {
-            return "";
+            Map<String, Object> claims = new HashMap<>();
```

Comment:
This code is almost identical to the one in `PravegaInterceptor` for `retrieveMasterToke`:

https://github.com/pravega/pravega/pull/2825/files#diff-bbac312563b334a3ae8acef0ac22365dR91

the only difference I could spot is the `claims`. Could you explain why splitting this way makes any difference with respect to escalation? If you have explained it somewhere else, please point me to it.

Reply:
There was no escalation in the present code.
But the API decided whether to create a master token or not based on the context. It is possible that any future code may inadvertently call this API without realizing the side effects. This is true about separating the code with a flag too.
The refactoring duplicates code but ensures that there is no such chance.